src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/JsonDictionaryConverter.cs

```diff
@@ -37,7 +37,28 @@ internal abstract class JsonDictionaryConverter<TDictionary, TKey, TValue> : Jso
         /// <summary>
         /// When overridden, create the collection. It may be a temporary collection or the final collection.
         /// </summary>
-        protected virtual void CreateCollection(ref Utf8JsonReader reader, ref ReadStack state) { }
+        protected virtual void CreateCollection(ref Utf8JsonReader reader, ref ReadStack state)
```

Comment:
this method is a bit confusing to me in general, why isn't it returning the constructed object and taking JsonTypeInfo as the arg

Reply:
In certain cases (such as immutable collections), the initial creation allocates an intermediate builder type (that isn't `TCollection` per se) so it would need to return `object`. We might be able to make this type safe by doing something like
```internal class JsonCollectionConverter<TCollection, TCollectionBuilder, TElement>
{
      public abstract TCollectionBuilder CreateCollection(...);
      public abstract TCollection ConvertCollection(..., TBuilder builder);
}
```
but I'm not sure it's worth the overengineering.